crates/edr_rpc_server/src/node.rs

```diff
@@ -437,4 +618,548 @@ mod tests {
+            .await
+            .unwrap();
+
+        let block1_hash = *fixture
+            .node
+            .block_and_total_difficulty_by_block_spec(&BlockSpec::Number(1))
+            .await
+            .unwrap()
+            .block
+            .hash();
+
+        let block_and_total_difficulty1 = fixture
+            .node
+            .block_and_total_difficulty_by_hash(&block1_hash)
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(
+            block_and_total_difficulty1.total_difficulty,
+            block_and_total_difficulty0.total_difficulty
+                + block_and_total_difficulty1.block.header().difficulty
+        );
+
+        assert_eq!(
+            block_and_total_difficulty1.total_difficulty,
+            block_and_total_difficulty0.total_difficulty
+                + block_and_total_difficulty1.block.header().difficulty
+        );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn block_transaction_count_by_hash() -> Result<()> {
+        let fixture = NodeTestFixture::new().await?;
+
+        let block_and_total_difficulty = fixture
+            .node
+            .block_and_total_difficulty_by_block_spec(&BlockSpec::Tag(BlockTag::Earliest))
+            .await
+            .unwrap();
+
+        let block_hash = block_and_total_difficulty.block.header().hash();
+
+        let count = fixture
+            .node
+            .block_transaction_count_by_hash(&block_hash)
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(count, 0);
+
+        let non_existing_count = fixture
+            .node
+            .block_transaction_count_by_hash(&B256::zero())
+            .await
+            .unwrap();
+
+        assert_eq!(non_existing_count, None);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn block_transaction_count_by_block_spec() -> Result<()> {
+        async fn assert_block_transaction_count(
+            fixture: &NodeTestFixture,
+            block_spec: BlockSpec,
+            expected_count: usize,
+        ) {
+            let count = fixture
+                .node
+                .block_transaction_count_by_block_spec(&block_spec)
+                .await
+                .unwrap();
+
+            assert_eq!(count, expected_count);
+        }
+
+        let fixture = NodeTestFixture::new().await?;
+
+        let block_and_total_difficulty = fixture
+            .node
+            .block_and_total_difficulty_by_block_spec(&BlockSpec::Tag(BlockTag::Earliest))
+            .await
+            .unwrap();
+
+        assert_block_transaction_count(&fixture, BlockSpec::Tag(BlockTag::Earliest), 0).await;
+        assert_block_transaction_count(&fixture, BlockSpec::Tag(BlockTag::Latest), 0).await;
+        assert_block_transaction_count(&fixture, BlockSpec::Tag(BlockTag::Pending), 0).await;
+        assert_block_transaction_count(&fixture, BlockSpec::Number(0), 0).await;
```

Comment:
I would create a unit test for the `block_by_block_spec` function that validates it works. That avoids having to create tests for all variants for every API that accepts a `BlockSpec`